File: day3/Exercise_day3/supplementary_material_omp/dot_prod.f90
!----------------------------------------------------------------------   
      program dot
!----------------------------------------------------------------------   

      use omp_lib

      implicit none

      integer      :: N = 100000000
      integer      :: i, num_threads
      real(8), dimension(:), allocatable  :: a, b
      real(8) :: time, sum, expected

!----------------------------------------------------------------------         
      allocate(a(N), b(N))

      num_threads = omp_get_max_threads()
      write(*,*) 'dot of vectors with length ', N, ' with ', num_threads, ' threads'

!.....init. arrays      
      do i=1,N
        a(i) = 1.0d0/2.0d0
        b(i) = i
      enddo
!----------------------------------------------------------------------   
!.....set timer      
      time = -omp_get_wtime();

!----------------------------------------------------------------------   
!.....loop to focus on
      sum = 0.0d0
      do i=1,n
        sum = sum + a(i) * b(i)
      end do
      
!.....get time      
      time = time + omp_get_wtime()

      expected = (N+1.0d0)*N/4.0d0;
      
      write(*,*) 'relative error ', abs(expected-sum)/expected
      write(*,*) 'took ', time, ' seconds'

      deallocate(a, b)

!----------------------------------------------------------------------   
      end program dot
!----------------------------------------------------------------------   

File: day3/Exercise_day3/supplementary_material_omp/dot_prod.cpp
#include <iostream>
#include <vector>

#include <omp.h>

int main(void){
    const int N = 100000000;
    std::vector<double> a(N);
    std::vector<double> b(N);

    int num_threads = omp_get_max_threads();
    std::cout << "dot of vectors with length " << N  << " with " << num_threads << " threads" << std::endl;

    // initialize the vectors
    for(int i=0; i<N; i++) {
        a[i] = 1./2.;
        b[i] = double(i+1);
    }

    double time = -omp_get_wtime();
    double dot=0.;

    for(int i=0; i<N; i++) {
        dot += a[i] * b[i];
    }
    time += omp_get_wtime();

    // use formula for sum of arithmetic sequence: sum(1:n) = (n+1)*n/2
    double expected = double(N+1)*double(N)/4.;
    std::cout << "dot product " << dot
              << (dot==expected ? " which matches the expected value "
                                : " which does not match the expected value ")
              << expected << std::endl;
    std::cout << "that took " << time << " seconds" << std::endl;
    return 0;
}

